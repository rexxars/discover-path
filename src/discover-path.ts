import {posix, win32} from 'path'
import nodeFs from 'fs'

export function discoverPathSync(target: string, options: DiscoverOptions = {}): string {
  const {readdirSync} = options.fs || nodeFs
  const {join} = getPathImpl(target)
  const paths = getPathSegments(target)

  let path
  let content
  let current
  const matchingSegments: string[] = [paths[0]]
  for (let i = 1; i < paths.length; i++) {
    path = join(...paths.slice(0, i))
    current = paths[i]
    content = readdirSync(path)

    // Direct match! (foo.jpg => foo.jpg)
    if (content.includes(current)) {
      matchingSegments.push(current)
      continue
    }

    // Case-insensitive match?
    const matches = getCaseInsensitiveMatches(current, content)

    // No matches?
    if (matches.length === 0) {
      throw new PathNotFoundError(target)
    }

    // One match? Auto-correct! (Foo.jpg => foo.jpg)
    if (matches.length === 1) {
      paths[i] = matches[0]
      matchingSegments.push(matches[0])
      continue
    }

    // More than one match? (Foo.jpg, foo.jpg, fOo.jpg)
    const head = matchingSegments.slice(0, i)
    const tail = paths.slice(i)
    const suggestions = tail.length === 1 ? matches.map((match) => join(...head, match)) : []
    throw new PathNotFoundError(target, suggestions)
  }

  return join(...matchingSegments)
}

function getCaseInsensitiveMatches(target: string, content: string[]): string[] {
  const lowercased = target.toLowerCase()
  const length = Math.min(content.length, 2)
  const matches = []
  for (let i = 0; i < length; i++) {
    if (lowercased === content[i].toLowerCase()) {
      matches.push(content[i])
    }
  }
  return matches
}

function getPathSegments(target: string): string[] {
  const {parse} = getPathImpl(target)
  const {root, dir, base} = parse(target)
  const path = getPathImpl(target)
  if (root === dir + base) {
    return [root]
  }

  const rootless = dir.slice(root.length).split(path.sep).filter(Boolean)
  return [root, ...rootless, base]
}

function getPathImpl(path: string) {
  return path[0] === '/' ? posix : win32
}

export interface MimimalFs {
  readdirSync(dir: string): string[]
}

export interface DiscoverOptions {
  fs?: MimimalFs
}

export class PathNotFoundError extends Error {
  path: string
  suggestions: string[]
  errno = -2
  code = 'ENOENT'
  syscall = 'scandir'

  constructor(missingPath: string, suggestions: string[] = []) {
    super(`ENOENT: no such file or directory, scandir '${missingPath}'`)

    this.path = missingPath
    this.suggestions = suggestions
    this.message +=
      suggestions.length === 0
        ? ''
        : `\nDid you mean:\n${suggestions.map((item) => `  - ${item}`).join('\n')}`
  }
}
