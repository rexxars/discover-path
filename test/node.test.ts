import {join, parse} from 'path'
import {discoverPathSync, PathNotFoundError} from '../src/discover-path'
import {getMockedFs} from './util/mockedFs'

describe('actual fs', () => {
  const fixturesPath = join(__dirname, 'fixtures')

  test('returns path verbatim on direct, case-sensitive match (file)', () => {
    const path = join(fixturesPath, 'foo', 'bAr', 'baz.txt')
    expect(discoverPathSync(path)).toEqual(path)
    expect(discoverPathSync(path, {})).toEqual(path)
  })

  test('returns path case-corrected on case-insensitive match (file)', () => {
    const path = join(fixturesPath, 'Foo', 'bar', 'Baz.txt')
    expect(discoverPathSync(path)).toEqual(join(fixturesPath, 'foo', 'bAr', 'baz.txt'))
    expect(discoverPathSync(path, {})).toEqual(join(fixturesPath, 'foo', 'bAr', 'baz.txt'))
  })

  test('returns path verbatim on direct, case-sensitive match (dir)', () => {
    const path = join(fixturesPath, 'foo', 'bAr')
    expect(discoverPathSync(path)).toEqual(path)
    expect(discoverPathSync(path, {})).toEqual(path)
  })

  test('returns path case-corrected on case-insensitive match (dir)', () => {
    const path = join(fixturesPath, 'Foo', 'bar')
    expect(discoverPathSync(path)).toEqual(join(fixturesPath, 'foo', 'bAr'))
    expect(discoverPathSync(path, {})).toEqual(join(fixturesPath, 'foo', 'bAr'))
  })

  test('returns path on root-lookup', () => {
    const {root} = parse(__dirname)
    expect(discoverPathSync(root)).toEqual(root)
    expect(discoverPathSync(root, {})).toEqual(root)
  })

  test('returns path on root-lookup', () => {
    const {root} = parse(__dirname)
    expect(discoverPathSync(root)).toEqual(root)
    expect(discoverPathSync(root, {})).toEqual(root)
  })

  test('throws on treating file as directory', () => {
    const path = join(fixturesPath, 'foo', 'bAr', 'baz.txt', 'yeah')
    expect(() => discoverPathSync(path)).toThrowError(/not a directory/)
    expect(() => discoverPathSync(path, {})).toThrowError(/not a directory/)
  })

  test('throws on no possible matches', () => {
    const path = join(fixturesPath, 'nein')
    expect(() => discoverPathSync(path)).toThrowError(/no such file or directory/)
    expect(() => discoverPathSync(path, {})).toThrowError(/no such file or directory/)
  })
})

describe('mocked posix', () => {
  test('returns path verbatim on direct, case-sensitive match (file)', () => {
    const path = '/foo/bar/baz.jpg'
    const fs = getMockedFs({'/': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual(path)
  })

  test('returns path case-corrected on case-insensitive match (file)', () => {
    const path = '/Foo/bar/Baz.jpg'
    const fs = getMockedFs({'/': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual('/foo/bar/baz.jpg')
  })

  test('returns path verbatim on direct, case-sensitive match (dir)', () => {
    const path = '/foo/bar'
    const fs = getMockedFs({'/': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual(path)
  })

  test('returns path case-corrected on case-insensitive match (dir)', () => {
    const path = '/Foo/bar'
    const fs = getMockedFs({'/': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual('/foo/bar')
  })

  test('returns path on root-lookup', () => {
    const fs = getMockedFs({'/': {foo: {}}})
    expect(discoverPathSync('/', {fs})).toEqual('/')
  })

  test('throws on treating file as directory', () => {
    const path = '/foo/bar/baz.jpg'
    const fs = getMockedFs({'/': {foo: true}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/not a directory/)
  })

  test('throws on no possible matches', () => {
    const path = '/foo/bar/baz.jpg'
    const fs = getMockedFs({'/': {foo: {}}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/no such file or directory/)
  })

  test('throw error with suggestions property on possible matches', () => {
    const path = '/foo/bar/baz'
    const fs = getMockedFs({'/': {foo: {bAr: {bAz: true, BAZ: true}}}})
    let error
    try {
      discoverPathSync(path, {fs})
    } catch (err) {
      error = err
    }

    expect(error).toBeInstanceOf(PathNotFoundError)
    expect(error.message).toMatchSnapshot()
    expect(error.suggestions).toEqual(['/foo/bAr/bAz', '/foo/bAr/BAZ'])
  })

  test('throws on no relevant matches', () => {
    const path = '/foo/bar/baz.jpg'
    const fs = getMockedFs({'/': {foo: {bap: true, folder: {}}}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/no such file or directory/)
  })

  test('throws on multiple matching paths (middle)', () => {
    const path = '/foo/bar/baz.jpg'
    const fs = getMockedFs({'/': {foo: {bAr: {'baz.jpg': true}, Bar: {'baz.jpg': true}}}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/no such file or directory/)
  })

  test('throws on multiple matching paths (end)', () => {
    const path = '/Users/rexxars/webdev/Discover-Path'
    const fs = getMockedFs({
      '/': {
        Users: {
          rexxars: {
            webdev: {
              'discover-path': {yep: true},
              'discover-Path': {yep: true},
            },
          },
        },
      },
    })
    expect(() => discoverPathSync(path, {fs})).toThrowErrorMatchingSnapshot()
  })
})

describe('mocked win32', () => {
  test('returns path verbatim on direct, case-sensitive match (file)', () => {
    const path = 'C:\\foo\\bar\\baz.jpg'
    const fs = getMockedFs({'C:\\': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual(path)
  })

  test('returns path case-corrected on case-insensitive match (file)', () => {
    const path = 'C:\\Foo\\bar\\Baz.jpg'
    const fs = getMockedFs({'C:\\': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual('C:\\foo\\bar\\baz.jpg')
  })

  test('returns path verbatim on direct, case-sensitive match (dir)', () => {
    const path = 'C:\\foo\\bar'
    const fs = getMockedFs({'C:\\': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual(path)
  })

  test('returns path case-corrected on case-insensitive match (dir)', () => {
    const path = 'C:\\Foo\\bar'
    const fs = getMockedFs({'C:\\': {foo: {bar: {'baz.jpg': true}}}})
    expect(discoverPathSync(path, {fs})).toEqual('C:\\foo\\bar')
  })

  test('returns path on root-lookup', () => {
    const fs = getMockedFs({'C:\\': {foo: {}}})
    expect(discoverPathSync('C:\\', {fs})).toEqual('C:\\')
  })

  test('throws on treating file as directory', () => {
    const path = 'C:\\foo\\bar\\baz.jpg'
    const fs = getMockedFs({'C:\\': {foo: true}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/not a directory/)
  })

  test('throws on no possible matches', () => {
    const path = 'C:\\foo\\bar\\baz.jpg'
    const fs = getMockedFs({'C:\\': {foo: {}}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/no such file or directory/)
  })

  test('throw error with suggestions property on possible matches', () => {
    const path = 'C:\\foo\\bar\\baz'
    const fs = getMockedFs({'C:\\': {foo: {bAr: {bAz: true, BAZ: true}}}})
    let error
    try {
      discoverPathSync(path, {fs})
    } catch (err) {
      error = err
    }

    expect(error).toBeInstanceOf(PathNotFoundError)
    expect(error.message).toMatchSnapshot()
    expect(error.suggestions).toEqual(['C:\\foo\\bAr\\bAz', 'C:\\foo\\bAr\\BAZ'])
  })

  test('throws on no relevant matches', () => {
    const path = 'C:\\foo\\bar\\baz.jpg'
    const fs = getMockedFs({'C:\\': {foo: {bap: true, folder: {}}}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/no such file or directory/)
  })

  test('throws on multiple matching paths (middle)', () => {
    const path = 'C:\\foo\\bar\\baz.jpg'
    const fs = getMockedFs({'C:\\': {foo: {bAr: {'baz.jpg': true}, Bar: {'baz.jpg': true}}}})
    expect(() => discoverPathSync(path, {fs})).toThrowError(/no such file or directory/)
  })

  test('throws on multiple matching paths (end)', () => {
    const path = 'C:\\Users\\rexxars\\webdev/Discover-Path'
    const fs = getMockedFs({
      '/': {
        Users: {
          rexxars: {
            webdev: {
              'discover-path': {yep: true},
              'discover-Path': {yep: true},
            },
          },
        },
      },
    })
    expect(() => discoverPathSync(path, {fs})).toThrowErrorMatchingSnapshot()
  })
})
