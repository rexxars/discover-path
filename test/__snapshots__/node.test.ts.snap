// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mocked posix throw error with suggestions property on possible matches 1`] = `
"ENOENT: no such file or directory, scandir '/foo/bar/baz'
Did you mean:
  - /foo/bAr/bAz
  - /foo/bAr/BAZ"
`;

exports[`mocked posix throws on multiple matching paths (end) 1`] = `
"ENOENT: no such file or directory, scandir '/Users/rexxars/webdev/Discover-Path'
Did you mean:
  - /Users/rexxars/webdev/discover-path
  - /Users/rexxars/webdev/discover-Path"
`;

exports[`mocked win32 throw error with suggestions property on possible matches 1`] = `
"ENOENT: no such file or directory, scandir 'C:\\\\foo\\\\bar\\\\baz'
Did you mean:
  - C:\\\\foo\\\\bAr\\\\bAz
  - C:\\\\foo\\\\bAr\\\\BAZ"
`;

exports[`mocked win32 throws on multiple matching paths (end) 1`] = `"ENOENT: no such file or directory, scandir 'C:\\\\'"`;
